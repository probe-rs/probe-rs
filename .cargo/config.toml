[alias]
xtask = "run --package xtask --"
probe-rs = "run --package probe-rs --features cli --bin probe-rs --"
smoke-tester = "run --package smoke_tester --"
