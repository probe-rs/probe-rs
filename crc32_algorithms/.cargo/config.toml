# Universal ARM binary configuration
# Only thumbv6m-none-eabi is needed now due to ARM upward compatibility

[target.thumbv6m-none-eabi]
rustflags = [
    "-C", "link-arg=-Tlink_minimal.x",
    "-C", "relocation-model=pic",
    "-C", "code-model=small",
    "-C", "panic=abort",
]