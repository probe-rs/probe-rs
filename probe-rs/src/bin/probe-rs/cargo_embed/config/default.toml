[default.probe]
# USB vendor ID
# usb_vid = "1337"
# USB product ID
# usb_pid = "1337"
# Serial number
# serial = "12345678"
# The protocol to be used for communicating with the target.
protocol = "Swd"
# The speed in kHz of the data link to the target.
# speed = 1337

[default.flashing]
# Whether or not the target should be flashed.
enabled = true
# Whether or not the target should be halted after reset.
# DEPRECATED, moved to reset section
halt_afterwards = false
# Whether or not bytes erased but not rewritten with data from the ELF
# should be restored with their contents before erasing.
restore_unwritten_bytes = false
# The path where an SVG of the assembled flash layout should be written to.
# flash_layout_output_path = "out.svg"
# Triggers a full chip erase instead of a page by page erase.
do_chip_erase = false

[default.reset]
# Whether or not the target should be reset.
# When flashing is enabled as well, the target will be reset after flashing.
enabled = true
# Whether or not the target should be halted after reset.
halt_afterwards = false

[default.general]
# The chip name of the chip to be debugged.
# chip = "name"
# A list of chip descriptions to be loaded during runtime.
chip_descriptions = []
# The default log level to be used. Possible values are one of:
#   "OFF", "ERROR", "WARN", "INFO", "DEBUG", "TRACE"
log_level = "WARN"
# Use this flag to assert the nreset & ntrst pins during attaching the probe to the chip.
connect_under_reset = false

[default.rtt]
# Whether or not an RTTUI should be opened after flashing.
enabled = false
# How the target handles RTT outputs that won't fit in the buffer.  This can be
# overridden per-channel. If left unset, the firmware will determine the default
# for each RTT up channel.
#   NoBlockSkip - Skip writing the data completely if it doesn't fit in its
#                 entirety.
#   NoBlockTrim - Write as much as possible of the data and ignore the rest.
#   BlockIfFull - Spin until the host reads data.  Can result in app freezing.
#
# up_mode = "BlockIfFull"

# A list of channel associations to be displayed. If left empty, all channels are displayed.
# up, down (Optional) - RTT channel numbers
# name     (Optional) - String to be displayed in the RTTUI tab
# up_mode  (Optional) - RTT channel specific as described above
# format   (Required) - How to interpret data from target firmware.  One of:
#              String - Directly show output from the target 
#              Defmt  - Format output on the host, see https://defmt.ferrous-systems.com/
#              BinaryLE - Display as raw hex
channels = [
    # { up = 0, down = 0, name = "name", up_mode = "BlockIfFull", format = "Defmt" },
]
# The duration in ms for which the logger should retry to attach to RTT.
timeout = 3000
# Whether timestamps in the RTTUI are enabled
show_timestamps = true
# Whether to save rtt history buffer on exit.
log_enabled = false
# Where to save rtt history buffer relative to manifest path.
log_path = "./logs"

[default.gdb]
# Whether or not a GDB server should be opened after flashing.
enabled = false
# The connection string in host:port format wher the GDB server will open a socket.
gdb_connection_string = "127.0.0.1:1337"
