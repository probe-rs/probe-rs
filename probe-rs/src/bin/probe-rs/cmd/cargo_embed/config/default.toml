[default.probe]
# USB vendor ID
# usb_vid = "1337"
# USB product ID
# usb_pid = "1337"
# Serial number
# serial = "12345678"
# The protocol to be used for communicating with the target.
protocol = "Swd"
# The speed in kHz of the data link to the target.
# speed = 1337

[default.flashing]
# Whether or not the target should be flashed.
enabled = true
# Whether or not the target should be halted after reset.
# DEPRECATED, moved to reset section
halt_afterwards = false
# Whether or not bytes erased but not rewritten with data from the ELF
# should be restored with their contents before erasing.
restore_unwritten_bytes = false
# The path where an SVG of the assembled flash layout should be written to.
# flash_layout_output_path = "out.svg"
# Triggers a full chip erase instead of a page by page erase.
do_chip_erase = false

[default.reset]
# Whether or not the target should be reset.
# When flashing is enabled as well, the target will be reset after flashing.
enabled = true
# Whether or not the target should be halted after reset.
halt_afterwards = false

[default.general]
# The chip name of the chip to be debugged.
# chip = "name"
# A list of chip descriptions to be loaded during runtime.
chip_descriptions = []
# The default log level to be used. Possible values are one of:
#   "OFF", "ERROR", "WARN", "INFO", "DEBUG", "TRACE"
log_level = "WARN"
# Use this flag to assert the nreset & ntrst pins during attaching the probe to the chip.
connect_under_reset = false
# Set a default binary like you would with --bin
# default_bin = "foo"

[default.rtt]
# Whether or not an RTTUI should be opened after flashing.
enabled = false
# The duration in ms for which the logger should retry to attach to RTT.
timeout = 3000
# Whether timestamps in the RTTUI are enabled
show_timestamps = true
# Whether locations for defmt messages are enabled in the RTTUI
show_location = false
# Whether to save rtt history buffer on exit.
log_enabled = false
# Where to save rtt history buffer relative to manifest path.
log_path = "./logs"
# A list of up (target -> host) channel settings associations to be displayed. If left empty, all channels are displayed.
# object key  - RTT channel identifier number
# mode     (Optional) - RTT operation mode. Describes how the target handles RTT outputs that won't
#                       fit in the buffer. If left unset, the firmware will determine the default
#                       for each RTT up channel.
#              * NoBlockSkip - Skip writing the data completely if it doesn't fit in its
#                            entirety.
#              * NoBlockTrim - Write as much as possible of the data and ignore the rest.
#              * BlockIfFull - Spin until the host reads data.  Can result in app freezing.
# format   (Optional) - How to interpret data from target firmware.  One of:
#              * String - Directly show output from the target (default)
#              * Defmt  - Format output on the host, see https://defmt.ferrous-systems.com/
#              * BinaryLE - Display as raw hex
# show_location (Optional) - Whether to show the location of defmt messages in the UI.
# show_timestamps (Optional) - Whether to show the timestamps of String messages in the UI.
# socket   (Optional) - Server socket address (for optional external frontend or endpoint).
up_channels = [
    # { channel = 0, mode = "BlockIfFull", format = "Defmt", show_location = true },
    # { channel = 1, mode = "BlockIfFull", format = "String", show_timestamps = false, socket = "127.0.0.1:12345" },
]

# A list of down (host -> target) channel settings. You can select a down channel for each UI tab,
# which will be used to send data to the target.
# object key  - RTT channel identifier number
# mode     (Optional) - RTT operation mode. Describes how the target handles RTT outputs that won't
#                       fit in the buffer. If left unset, the firmware will determine the default
#                       for each RTT down channel.
down_channels = [
    # { channel = 0, mode = "BlockIfFull" }
]

# UI tab settings. All up channels are displayed, except when hidden here. You can specify how each
# tab is displayed and whether they allow sending data to the target.
# up_channel              - The channel_number of the RTT up channel to display
# hide         (Optional) - Whether to hide the tab. Defaults to false.
# down_channel (Optional) - The channel_number of the RTT down channel to use for this tab.
# name         (Optional) - String to be displayed in the RTTUI tab. Defaults to the channel name.
tabs = [
    # { up_channel = 0, down_channel = 0, name = "My channel" },
    # { up_channel = 1, hide = true },
]

[default.gdb]
# Whether or not a GDB server should be opened after flashing.
enabled = false
# The connection string in host:port format wher the GDB server will open a socket.
gdb_connection_string = "127.0.0.1:1337"
