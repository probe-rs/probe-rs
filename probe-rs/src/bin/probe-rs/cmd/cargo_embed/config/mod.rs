use anyhow::bail;
use figment::{
    providers::{Format, Json, Toml, Yaml},
    Figment,
};
use probe_rs::probe::WireProtocol;
use probe_rs::rtt::ChannelMode;
use serde::{Deserialize, Serialize};
use std::{net::SocketAddr, path::PathBuf, time::Duration};

use crate::util::{logging::LevelFilter, rtt::DataFormat};

use super::rttui::tab::TabConfig;

/// A struct which holds all configs.
#[derive(Debug, Clone)]
pub struct Configs {
    figment: Figment,
}

/// The main struct holding all the possible config options.
#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct Config {
    pub general: General,
    pub flashing: Flashing,
    pub reset: Reset,
    pub probe: Probe,
    pub rtt: Rtt,
    pub gdb: Gdb,
}

/// The probe config struct holding all the possible probe options.
#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct Probe {
    pub usb_vid: Option<String>,
    pub usb_pid: Option<String>,
    pub serial: Option<String>,
    pub protocol: WireProtocol,
    pub speed: Option<u32>,
}

/// The flashing config struct holding all the possible flashing options.
#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct Flashing {
    pub enabled: bool,
    #[deprecated(
        since = "0.9.0",
        note = "The 'halt_afterwards' key has moved to the 'reset' section"
    )]
    pub halt_afterwards: bool,
    pub restore_unwritten_bytes: bool,
    pub flash_layout_output_path: Option<String>,
    pub do_chip_erase: bool,
}

/// The reset config struct holding all the possible reset options.
#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct Reset {
    pub enabled: bool,
    pub halt_afterwards: bool,
}

/// The general config struct holding all the possible general options.
#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct General {
    pub chip: Option<String>,
    pub chip_descriptions: Vec<String>,
    pub log_level: Option<LevelFilter>,
    pub derives: Option<String>,
    /// Use this flag to assert the nreset & ntrst pins during attaching the probe to the chip.
    pub connect_under_reset: bool,
    pub default_bin: Option<String>,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct UpChannelConfig {
    pub channel: usize,
    #[serde(default)]
    pub mode: Option<ChannelMode>,
    #[serde(default)]
    pub format: DataFormat,
    #[serde(default)]
    pub show_location: Option<bool>,
    #[serde(default)]
    pub socket: Option<SocketAddr>,
    // TODO: it should be possible to move these into DataFormat
    #[serde(default)]
    /// Control the inclusion of timestamps for DataFormat::String.
    pub show_timestamps: Option<bool>,
    #[serde(default)]
    /// Control the output format for DataFormat::Defmt.
    pub defmt_log_format: Option<String>,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct DownChannelConfig {
    pub channel: usize,
    #[serde(default)]
    pub mode: Option<ChannelMode>,
}

/// The rtt config struct holding all the possible rtt options.
#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct Rtt {
    pub enabled: bool,
    /// Up mode, when not specified per-channel.  Target picks if neither is set
    pub up_mode: Option<ChannelMode>,
    /// Channels to be displayed, and options for them
    pub up_channels: Vec<UpChannelConfig>,
    /// Channels to be displayed, and options for them
    pub down_channels: Vec<DownChannelConfig>,
    /// UI tab configuration
    pub tabs: Vec<TabConfig>,
    /// Connection timeout in ms.
    #[serde(with = "duration_ms")]
    pub timeout: Duration,
    /// Whether to show timestamps in RTTUI
    pub show_timestamps: bool,
    /// Whether to show location info in RTTUI for defmt channels.
    pub show_location: bool,
    /// Whether to save rtt history buffer on exit to file named history.txt
    pub log_enabled: bool,
    /// Where to save rtt history buffer relative to manifest path.
    pub log_path: PathBuf,
}

impl Rtt {
    /// Returns the configuration for the specified up channel number, if it exists.
    pub fn up_channel_config(&self, channel_number: usize) -> Option<&UpChannelConfig> {
        self.up_channels
            .iter()
            .find(|ch| ch.channel == channel_number)
    }
}

mod duration_ms {
    use std::time::Duration;

    use serde::{Deserialize, Deserializer, Serializer};

    pub fn serialize<S>(duration: &Duration, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_u128(duration.as_millis())
    }

    pub fn deserialize<'de, D>(deserializer: D) -> Result<Duration, D::Error>
    where
        D: Deserializer<'de>,
    {
        u64::deserialize(deserializer).map(Duration::from_millis)
    }
}

/// The gdb config struct holding all the possible gdb options.
#[derive(Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub struct Gdb {
    pub enabled: bool,
    pub gdb_connection_string: Option<String>,
}

impl Configs {
    pub fn new(conf_dir: PathBuf) -> Configs {
        // Start off by merging in the default configuration file.
        let mut figments =
            Figment::new().merge(Toml::string(include_str!("default.toml")).nested());

        // Ordered list of config files, which are handled in the order specified here.
        let config_files = [
            // The following files are intended to be project-specific and would normally be
            // included in a project's source repository.
            ".embed",
            "Embed",
            // The following files are intended to hold personal or exceptional settings, which
            // are not useful for other users, and would NOT normally be included in a project's
            // source repository.
            ".embed.local",
            "Embed.local",
        ];

        for file in &config_files {
            let mut toml_path: std::path::PathBuf = conf_dir.clone();
            toml_path.push(format!("{file}.toml"));

            let mut json_path = conf_dir.clone();
            json_path.push(format!("{file}.json"));

            let mut yaml_path = conf_dir.clone();
            yaml_path.push(format!("{file}.yaml"));

            let mut yml_path = conf_dir.clone();
            yml_path.push(format!("{file}.yml"));

            figments = Figment::from(figments)
                .merge(Toml::file(toml_path).nested())
                .merge(Json::file(json_path).nested())
                .merge(Yaml::file(yaml_path).nested())
                .merge(Yaml::file(yml_path).nested());
        }
        Configs { figment: figments }
    }

    pub fn prof_names(&self) -> Vec<String> {
        self.figment
            .profiles()
            .map(|p| String::from(p.as_str().as_str()))
            .collect()
    }

    /// Extract the requested config, but only if the profile has been explicitly defined in the
    /// configuration files etc. (selecting an arbitrary undefined profile with Figment will coerce
    /// it into existence - inheriting from the default config).
    pub fn select_defined(self: Configs, name: &str) -> anyhow::Result<Config> {
        let defined_profiles = self.prof_names();
        let requested_profile_defined: bool = defined_profiles
            .iter()
            .any(|p| p.to_lowercase() == name.to_lowercase());

        let figext: figment::error::Result<Config> = self.figment.select(name).extract();
        match figext {
            Err(figerr) => {
                // Join all the figment errors into a multiline string.
                bail!(
                    "Failed to parse supplied configuration:\n{}",
                    figerr
                        .into_iter()
                        .map(|e| e.to_string())
                        .collect::<Vec<String>>()
                        .join("\n")
                );
            }
            Ok(config) => {
                // Gross errors (e.g. config file syntax problems) have already been caught by the
                // other match arm.  Guard against Figment coercing previously undefined profiles
                // into existance.
                if !requested_profile_defined {
                    bail!(
                        "the requested configuration profile \"{}\" hasn't been defined (defined profiles: {})",
                        name,
                        defined_profiles.join(", ")
                    );
                }
                Ok(config)
            }
        }
    }
    #[cfg(test)]
    pub fn new_with_test_data(conf_dir: PathBuf) -> Configs {
        let mut cfs = Configs::new(conf_dir);
        cfs.figment = cfs.figment.merge(
            Toml::string(
                r#"
            [default]
               bogusInvalidItem = "oops"
               "#,
            )
            .nested(),
        );
        cfs
    }
}

#[cfg(test)]
mod test {
    use super::Configs;

    #[test]
    fn default_profile() {
        // Ensure the default config can be parsed.
        let configs = Configs::new(std::env::current_dir().unwrap());
        let _config = configs.select_defined("default").unwrap();
    }
    #[test]
    fn non_existant_profile_is_error() {
        // Selecting a non-existant profile.
        let configs = Configs::new(std::env::current_dir().unwrap());
        let _noprofile = configs.select_defined("nonexxistantprofle").unwrap_err();
    }
    #[test]
    fn unknown_config_items_in_config_fail() {
        // Selecting a profile with an invalid item.
        let configs = Configs::new_with_test_data(std::env::current_dir().unwrap());
        let _superfluous: anyhow::Error = configs.select_defined("default").unwrap_err();
    }
}
