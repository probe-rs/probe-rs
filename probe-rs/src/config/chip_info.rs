use crate::architecture::arm::ArmChipInfo;

/// Information about a chip which is used
/// for automatic detection of the connected chip.
///
/// For ARM-based chips, the function [`ArmProbeInterface::read_chip_info_from_rom_table()`][r] is
/// used to read the information from the target.
///
/// [r]: crate::architecture::arm::ArmProbeInterface::read_chip_info_from_rom_table
#[derive(Debug)]
pub(crate) enum ChipInfo {
    /// ARM specific information for chip
    /// auto-detection. See [ArmChipInfo].
    Arm(ArmChipInfo),
}

impl From<ArmChipInfo> for ChipInfo {
    fn from(info: ArmChipInfo) -> Self {
        ChipInfo::Arm(info)
    }
}
