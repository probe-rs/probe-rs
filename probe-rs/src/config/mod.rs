#![warn(missing_docs)]

//! Target specific configuration
//!
//! For debugging and flashing different chips, called *target* in probe-rs, some
//! target specific configuration is required. This includes the architecture of
//! the chip, e.g. RISC-V or ARM, and information about the memory map of the target,
//! which can be used together with a flash algorithm to program the flash memory
//! of a target.
//!
//! ## Built-in targets
//!
//! The built-in targets are added at build-time, from the `build.rs` script.
//! They are generated from the target files in the `targets/` subfolder of this
//! crate.
//!
//! The built-in targets can be disabled by not including the `builtin-targets` feature.
//!
//! ## Adding targets at runtime
//!
//! To add a target at runtime, the [add_target_from_yaml] function can
//! be used to read targets from a YAML file.
//!

mod chip_info;
mod registry;
pub mod sequences;
mod target;

pub use probe_rs_target::{
    Chip, ChipFamily, Core, CoreType, FlashProperties, GenericRegion, InstructionSet, MemoryRange,
    MemoryRegion, NvmRegion, PageInfo, RamRegion, RawFlashAlgorithm, ScanChainElement,
    SectorDescription, SectorInfo, TargetDescriptionSource,
};

pub use registry::{
    add_target_from_yaml, families, get_target_and_family_by_name, get_target_by_name,
    get_targets_by_family_name, search_chips, RegistryError,
};
pub use target::{DebugSequence, Target, TargetParseError, TargetSelector};

// Crate-internal API
pub(crate) use chip_info::ChipInfo;
pub(crate) use registry::get_target_by_chip_info;
pub(crate) use target::CoreExt;
