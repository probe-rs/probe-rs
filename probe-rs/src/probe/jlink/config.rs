#[derive(Default, Clone, Copy, Debug)]
#[allow(dead_code)]
pub struct JlinkConfig {
    pub usb_address: Option<u8>,
    pub kickstart_power: Option<bool>,
    pub ip_address: Option<[u8; 4]>,
    pub subnet_mask: Option<[u8; 4]>,
    pub mac_address: Option<[u8; 6]>,
}

impl JlinkConfig {
    pub fn parse(data: [u8; 256]) -> Result<Self, String> {
        let usb_address = match data[0] {
            0 => Some(0),
            1 => Some(1),
            2 => Some(2),
            0xFF => None,
            other => return Err(format!("Unexpected USB address configured: {other}")),
        };

        let kickstart_power = match u32::from_le_bytes([data[4], data[5], data[6], data[7]]) {
            0 => Some(false),
            1 => Some(true),
            u32::MAX => None,
            other => return Err(format!("Unexpected kickstart power value: {other:#010x}")),
        };

        let ip_address = match data[32..36] {
            [0xFF, 0xFF, 0xFF, 0xFF] => None,
            [a, b, c, d] => Some([a, b, c, d]),
            _ => unreachable!(),
        };

        let subnet_mask = match data[36..40] {
            [0xFF, 0xFF, 0xFF, 0xFF] => None,
            [a, b, c, d] => Some([a, b, c, d]),
            _ => unreachable!(),
        };

        let mac_address = match data[48..54] {
            [0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF] => None,
            [a, b, c, d, e, f] => Some([a, b, c, d, e, f]),
            _ => unreachable!(),
        };

        Ok(Self {
            usb_address,
            kickstart_power,
            ip_address,
            subnet_mask,
            mac_address,
        })
    }
}
