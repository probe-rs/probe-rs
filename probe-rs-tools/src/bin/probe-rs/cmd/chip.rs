use bytesize::ByteSize;
use probe_rs::config::MemoryRegion;

#[derive(clap::Parser)]
pub struct Cmd {
    #[clap(subcommand)]
    subcommand: Subcommand,
}

#[derive(clap::Subcommand)]
/// Inspect internal registry of supported chips
enum Subcommand {
    /// Lists all the available families and their chips with their full.
    #[clap(name = "list")]
    List,
    /// Shows chip properties of a specific chip
    #[clap(name = "info")]
    Info {
        /// The name of the chip to display.
        name: String,
    },
}

impl Cmd {
    pub fn run(self) -> anyhow::Result<()> {
        let output = std::io::stdout().lock();

        match self.subcommand {
            Subcommand::List => print_families(output),
            Subcommand::Info { name } => print_chip_info(output, &name),
        }
    }
}

/// Print all the available families and their contained chips to the
/// commandline.
pub fn print_families(mut output: impl std::io::Write) -> anyhow::Result<()> {
    writeln!(output, "Available chips:")?;
    for family in probe_rs::config::families() {
        writeln!(output, "{}", &family.name)?;
        writeln!(output, "    Variants:")?;
        for variant in family.variants() {
            writeln!(output, "        {}", variant.name)?;
        }
    }
    Ok(())
}

/// Print all the available families and their contained chips to the
/// commandline.
pub fn print_chip_info(mut output: impl std::io::Write, name: &str) -> anyhow::Result<()> {
    writeln!(output, "{}", name)?;
    let target = probe_rs::config::get_target_by_name(name)?;
    writeln!(output, "Cores ({}):", target.cores.len())?;
    for core in target.cores {
        writeln!(
            output,
            "    - {} ({:?})",
            core.name.to_ascii_lowercase(),
            core.core_type
        )?;
    }

    fn get_range_len(range: &std::ops::Range<u64>) -> u64 {
        range.end - range.start
    }

    for memory in target.memory_map {
        let range = memory.address_range();
        let size = ByteSize(get_range_len(&range)).to_string_as(true);
        let kind = match memory {
            MemoryRegion::Ram(_) => "RAM",
            MemoryRegion::Generic(_) => "Generic",
            MemoryRegion::Nvm(_) => "NVM",
        };
        writeln!(output, "{kind}: {range:#010x?} ({size})")?
    }
    Ok(())
}

#[test]
fn single_chip_output() {
    let mut buff = Vec::new();
    print_chip_info(&mut buff, "nrf52840_xxaa").unwrap();

    // output should be valid utf8
    let output = String::from_utf8(buff).unwrap();

    insta::assert_snapshot!(output);
}

#[test]
fn multiple_chip_output() {
    let mut buff = Vec::new();
    let error = print_chip_info(&mut buff, "nrf52").unwrap_err();

    insta::assert_snapshot!(error.to_string());
}
