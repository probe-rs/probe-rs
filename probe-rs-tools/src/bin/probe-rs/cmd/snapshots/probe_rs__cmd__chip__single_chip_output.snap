---
source: probe-rs-tools/src/bin/probe-rs/cmd/chip.rs
expression: output
---
nrf52840_xxaa
Cores (1):
    - main (Armv7em)
RAM: 0x20000000..0x20040000 (256.0 kiB)
RAM: 0x00800000..0x00840000 (256.0 kiB)
NVM: 0x00000000..0x00100000 (1.0 MiB)
NVM: 0x10001000..0x10002000 (4.0 kiB)
